core/src/alpenglow_consensus/voting_loop.rs

```diff
@@ -830,53 +846,95 @@ impl VotingLoop {
         // replay_timing.generate_vote_us += generate_time.as_us();
-        let GenerateVoteTxResult::Tx(vote_tx) = vote_tx_result else {
-            return;
+        let GenerateVoteTxResult::BLSMessage(bls_message) = vote_tx_result else {
+            return false;
         };
 
-        if let Err(e) = Self::add_vote_and_maybe_update_commitment(
+        if let Err(e) = Self::add_message_and_maybe_update_commitment(
             &context.identity_keypair.pubkey(),
-            &vote,
-            &context.vote_account_pubkey,
-            vote_tx.clone().into(),
+            &bls_message,
             cert_pool,
             &context.commitment_sender,
         ) {
             if !is_refresh {
                 warn!("Unable to push our own vote into the pool {}", e);
-                return;
+                return false;
             }
         };
 
-        // Update and save the vote history
-        if !is_refresh {
-            context.vote_history.add_vote(vote);
-        }
-        let saved_vote_history =
-            SavedVoteHistory::new(&context.vote_history, &context.identity_keypair).unwrap_or_else(
-                |err| {
-                    error!("Unable to create saved vote history: {:?}", err);
-                    std::process::exit(1);
-                },
-            );
-
         // Send the vote over the wire
+        let saved_vote_history = if is_refresh {
+            None
+        } else {
+            // Update and save the vote history
+            context.vote_history.add_vote(vote);
+            let saved_vote_history =
+                SavedVoteHistory::new(&context.vote_history, &context.identity_keypair)
+                    .unwrap_or_else(|err| {
+                        error!("Unable to create saved vote history: {:?}", err);
+                        std::process::exit(1);
+                    });
+
+            Some(SavedVoteHistoryVersions::from(saved_vote_history))
+        };
         context
             .voting_sender
-            .send(VoteOp::PushAlpenglowVote {
-                tx: vote_tx,
-                slot: bank.slot(),
-                saved_vote_history: SavedVoteHistoryVersions::from(saved_vote_history),
+            .send(VoteOp::PushAlpenglowBLSMessage {
+                bls_message,
+                slot: vote.slot(),
+                saved_vote_history,
             })
             .unwrap_or_else(|err| warn!("Error: {:?}", err));
+        true
+    }
+
+    fn get_bls_keypair(
+        context: &mut VotingContext,
+        authorized_voter_keypair: &Arc<Keypair>,
+    ) -> Result<Arc<BLSKeypair>, BlsError> {
+        let pubkey = authorized_voter_keypair.pubkey();
+        if let Some(existing) = context.derived_bls_keypairs.get(&pubkey) {
+            return Ok(existing.clone());
+        }
+
+        let bls_keypair = Arc::new(BLSKeypair::derive_from_signer(
+            authorized_voter_keypair,
+            b"alpenglow",
```

Comment:
nit: let's make this a constant in the top level alpenglow module

Reply:
Done.